flocker/control/_persistence.py

```diff
@@ -21,16 +23,151 @@
 
 
 # Serialization marker storing the class name:
 _CLASS_MARKER = u"$__class__$"
 
+# The latest configuration version. Configuration versions are
+# always integers. Requires a corresponding Configuration_Vx class,
+# where x is the _CONFIG_VERSION.
+_CONFIG_VERSION = 1
+
+# Map of serializable class names to classes
+_CONFIG_CLASS_MAP = {cls.__name__: cls for cls in SERIALIZABLE_CLASSES}
+
+
+class ConfigurationMigrationError(Exception):
+    """
+    Error raised when a configuration migration is unable to take place
+    or complete successfully.
+    """
+
+
+def migrate_configuration(source_version, target_version, config):
+    """
+    Migrate a persisted configuration from one version to another
+    in sequential upgrades, e.g. a source version of 1 and target
+    version of 3 will perform two upgrades, from version 1 to 2,
+    followed by 2 to 3.
+
+    Calls the correct ``ConfigurationMigration`` class methods for
+    the suppled source and target versions.
+
+    :param int source_version: The version to migrate from.
+    :param int target_version: The version to migrate to.
+    :param bytes config: The JSON-encoded source configuration.
+
+    :return bytes: The updated JSON configuration after migration.
+    """
+    upgraded_config = config
+    current_version = source_version
+    for upgrade_version in range(source_version + 1, target_version + 1):
+        migration_method = (
+            u"configuration_v%d_v%d"
+            % (current_version, upgrade_version)
+        )
+        try:
+            migration = getattr(ConfigurationMigration, migration_method)
+        except AttributeError:
+            message = (
+                u"Unable to find a migration path for a version " +
+                unicode(source_version) + u" to version " +
+                unicode(target_version) + u" configuration. " +
+                u"No migration method exists for v" +
+                unicode(current_version) + u" to v" +
+                unicode(upgrade_version) + u"."
+            )
+            raise ConfigurationMigrationError(message)
+        upgraded_config = migration(config)
+        current_version = current_version + 1
+    return upgraded_config
+
+
+class IConfiguration(Interface):
+    """
+    An ``IConfiguration`` implementation provides a serializer and
+    deserializer for a ``Configuration`` model.
+    """
+    def serialize(config):
+        """
+        Serialize the supplied configuration model to JSON.
+
+        :param Configuration config: The configuration to serialize.
+        :return bytes: The JSON representation.
+        """
+
+    def deserialize(config):
+        """
+        Deserialize the supplied JSON to a ``Configuration`` model.
+
+        :param bytes config: The JSON configuration to deserialize.
+        :return Configuration: The configuration model.
+        """
+
 
-class _ConfigurationEncoder(JSONEncoder):
+@implementer(IConfiguration)
+class Configuration_V1(object):
```

Comment:
In real code you wouldn't typically keep around the old IConfiguration classes since they may not work at all, no? So Configuration_V1 ought not exist.